cloudant-client/src/test/java/com/cloudant/tests/HttpIamTest.java

```diff
@@ -0,0 +1,434 @@
+import static com.cloudant.tests.util.MockWebServerResources.OK_IAM_COOKIE_2;
+import static com.cloudant.tests.util.MockWebServerResources.iamSession;
+import static com.cloudant.tests.util.MockWebServerResources.iamSessionUnquoted;
+import static com.cloudant.tests.util.MockWebServerResources.setCookie;
+import static org.hamcrest.CoreMatchers.anyOf;
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+import com.cloudant.client.api.CloudantClient;
+import com.cloudant.client.org.lightcouch.CouchDbException;
+import com.cloudant.http.Http;
+
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+
+import mockit.Invocation;
+import mockit.Mock;
+import mockit.MockUp;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+
+import java.nio.charset.Charset;
+
+/**
+ * Created by tomblench on 30/06/2017.
+ */
+
+public class HttpIamTest {
+
+    @Rule
+    public MockWebServer mockWebServer = new MockWebServer();
+
+    @Rule
+    public MockWebServer mockIamServer = new MockWebServer();
+
+    final static String hello = "{\"hello\":\"world\"}\r\n";
+    final static String iamApiKey = "iam";
+    final static String iamTokenEndpoint = "/oidc/token";
+
+    /**
+     * Mock System.getProperty to fake the value of the IAM server
+     *
+     * Do this instead of setting the property in the real System because it pollutes the JVM
+     */
+    class SystemMock extends MockUp<System> {
+
+        @Mock
+        public String getProperty(Invocation inv, String key) {
+            if (key.equals("com.cloudant.client.iamserver")) {
+                return mockIamServer.url(iamTokenEndpoint).toString();
+            }
+            return inv.proceed(key);
+        }
+
+        @Mock
+        public String getProperty(Invocation inv, String key, String def) {
+            if (key.equals("com.cloudant.client.iamserver")) {
+                return mockIamServer.url(iamTokenEndpoint).toString();
+            }
+            return inv.proceed(key, def);
+        }
+    }
+
+    @Before
+    public void setupSystemMock() {
+        new SystemMock();
+    }
+
+    /**
+     * Test IAM token and cookie flow:
+     * - GET a resource on the cloudant server
+     * - Cookie jar empty, so get IAM token followed by session cookie
+     * - GET now proceeds as normal, expected cookie value is sent in header
+     * @throws Exception
+     */
+    @Test
+    public void iamTokenAndCookieSuccessful() throws Exception {
+
+        // Request sequence
+        // _session request to get Cookie
+        // GET request -> 200 with a Set-Cookie
+        mockWebServer.enqueue(OK_IAM_COOKIE);
+        mockWebServer.enqueue(new MockResponse().setResponseCode(200).addHeader("Set-Cookie",
```

Comment:
I don't think there will be a `Set-Cookie` here, I think this will just a be a normal response, the `Set-Cookie` has already happened on the previous `_iam_session` response.